src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/AnyOS/ManagedPal.Decrypt.cs

```diff
@@ -0,0 +1,158 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.Security.Cryptography;
+using System.Security.Cryptography.Asn1;
+using System.Security.Cryptography.Pkcs;
+using System.Security.Cryptography.Pkcs.Asn1;
+using System.Security.Cryptography.X509Certificates;
+
+namespace Internal.Cryptography.Pal.AnyOS
+{
+    internal sealed partial class ManagedPkcsPal : PkcsPal
+    {
+        private sealed class ManagedDecryptorPal : DecryptorPal
+        {
+            private byte[] _dataCopy;
+            private EnvelopedDataAsn _envelopedData;
+
+            public ManagedDecryptorPal(
+                byte[] dataCopy,
+                EnvelopedDataAsn envelopedDataAsn,
+                RecipientInfoCollection recipientInfos)
+                : base(recipientInfos)
+            {
+                _dataCopy = dataCopy;
+                _envelopedData = envelopedDataAsn;
+            }
+
+            public override ContentInfo TryDecrypt(
+                RecipientInfo recipientInfo,
+                X509Certificate2 cert,
+                X509Certificate2Collection originatorCerts,
+                X509Certificate2Collection extraStore,
+                out Exception exception)
+            {
+                // When encryptedContent is null Windows seems to decrypt the CEK first,
+                // then return a 0 byte answer.
+
+                byte[] cek;
```

Comment:
We should pin _cek_ to minimize its exposure to GC relocation. It should also be cleared when this method returns. (Ideally we'd just stackalloc it, but `alg.CreateDecryptor` makes that a bit difficult.)

Same comment for encrypt routine.